[ML] Add option for anomaly charts for metric detector should plot min, mean or max as appropriate

qn895 opened the pull request:
## Summary

This PR addresses https://github.com/elastic/kibana/issues/65078 by:
1) Plot the function associated with the record with the highest anomaly high score in the Single Metric Viewer and Anomaly Explorer by default.
![ae_metric_previews](https://user-images.githubusercontent.com/43350163/97215169-1e8aa980-1792-11eb-8fce-38b4cf4adb07.gif)
![smv_metric_preview_1](https://user-images.githubusercontent.com/43350163/97215390-5c87cd80-1792-11eb-90e7-746dfbe909f7.gif)

2) Add additional select options so users can choose what function want to view by (e.g. `avg`, `min`, or `max`) where the anomaly markers will only show up if they correspond to the right view.
![smv_metric_preview_2](https://user-images.githubusercontent.com/43350163/97215732-d750e880-1792-11eb-82a7-fab642e4940b.gif)


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/)
* Commit: 3c8972a6016d952a4380f5113ff28f50cb7f16ff

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### async chunks size
| id | [before](https://github.com/elastic/kibana/commit/3f912547b89e3aa9df9d23407e3f99865142204d) | [after](https://github.com/elastic/kibana/commit/3c8972a6016d952a4380f5113ff28f50cb7f16ff) | diff |
| --- | --- | --- | --- |
| `ml` | 7.3MB | 7.3MB | +10.3KB |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/","number":"83747","commit":"3c8972a6016d952a4380f5113ff28f50cb7f16ff"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

darnautov commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer.js:
```diff
@@ -826,6 +860,36 @@ export class TimeSeriesExplorer extends React.Component {
+          }
+        })
+        .catch((resp) => {
+          console.log(
+            'Time series explorer - error getting record with highest anomaly score:',
+            resp
+          );
```
instead of logging, we should probably show this error to the user with a toast notification

qn895 replied:
Updated to show toast notification here https://github.com/elastic/kibana/pull/81662/commits/19e20ee5f7a61327b729440a84a8fdf6e87a6a4a

darnautov commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer.js:
```diff
@@ -1186,6 +1253,42 @@ export class TimeSeriesExplorer extends React.Component {
                 />
               );
             })}
+            {actualPlotFunction && (
+              <EuiFlexItem style={{ textAlign: 'right' }}>
+                <EuiFormRow
+                  label={i18n.translate('xpack.ml.timeSeriesExplorer.metricPlotByOption', {
+                    defaultMessage: 'Function',
+                  })}
+                >
+                  <EuiSelect
+                    options={[
+                      {
+                        value: 'avg',
+                        text: i18n.translate('xpack.ml.timeSeriesExplorer.plotByAvgOptionLabel', {
+                          defaultMessage: 'average',
+                        }),
+                      },
+                      {
+                        value: 'min',
+                        text: i18n.translate('xpack.ml.timeSeriesExplorer.plotByMinOptionLabel', {
+                          defaultMessage: 'min',
+                        }),
+                      },
+                      {
+                        value: 'max',
+                        text: i18n.translate('xpack.ml.timeSeriesExplorer.plotByMaxOptionLabel', {
+                          defaultMessage: 'max',
+                        }),
+                      },
+                    ]}
+                    value={actualPlotFunction ?? 'avg'}
+                    onChange={(e) => this.setPlotByFunction(e.target.value)}
+                    aria-label="Pick function to plot by (min, max, or average) if metric function"
+                  />
+                </EuiFormRow>
+              </EuiFlexItem>
+            )}
```
Timeseriesexplorer component is already huge, consider moving this part.

qn895 replied:
Moved to a separate file here https://github.com/elastic/kibana/pull/81662/commits/21bb6cb31d31959d096b318b15f2426c840d6ac9

darnautov commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer.js:
```diff
@@ -826,6 +860,36 @@ export class TimeSeriesExplorer extends React.Component {
     if (detectorId !== selectedDetectorIndex) {
       appStateHandler(APP_STATE_ACTION.SET_DETECTOR_INDEX, detectorId);
     }
+    // if the detector's function is metric, fetch the highest scoring anomaly record
```
I guess this logic can be moved outside of the TimeSeriesExplorer component

qn895 replied:
Refactored here https://github.com/elastic/kibana/pull/81662/commits/19e20ee5f7a61327b729440a84a8fdf6e87a6a4a

darnautov commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer.js:
```diff
@@ -421,6 +430,27 @@ export class TimeSeriesExplorer extends React.Component {
       );
   };
 
+  loadPlotByOptions = async () => {
```
Consider moving it outside of TimeSeriesExplorer

qn895 replied:
Moved the logic into a separate file here https://github.com/elastic/kibana/pull/81662/commits/19e20ee5f7a61327b729440a84a8fdf6e87a6a4a

peteharverson commented:
As discussed, if feasible, I think it makes sense to filter the table and mini swim lane anomalies for the detector function (min / avg / max) as well as the main chart. That way the view shows information for the single series (detector, partitioning fields, function) being viewed.

![image](https://user-images.githubusercontent.com/7405507/97477175-2d668d00-1947-11eb-87f5-ccb05ee9d7bc.png)

peteharverson commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer.js:
```diff
@@ -1186,6 +1253,42 @@ export class TimeSeriesExplorer extends React.Component {
+                    defaultMessage: 'Function',
+                  })}
+                >
+                  <EuiSelect
```
Is it possible to decrease the width on this element?

![image](https://user-images.githubusercontent.com/7405507/97477681-d3b29280-1947-11eb-90b1-b52749d324a2.png)

qn895 replied:
Decreased the width here  https://github.com/elastic/kibana/pull/81662/commits/21bb6cb31d31959d096b318b15f2426c840d6ac9

<img width="734" alt="Screen Shot 2020-11-03 at 14 09 12" src="https://user-images.githubusercontent.com/43350163/98035510-3d1a2180-1dde-11eb-8f6f-5b15246a041c.png">

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85403/)
* Commit: 21bb6cb31d31959d096b318b15f2426c840d6ac9

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [before](https://github.com/elastic/kibana/commit/6f2dc107275b8a2d264cb9148c70302c7d2d8275) | [after](https://github.com/elastic/kibana/commit/21bb6cb31d31959d096b318b15f2426c840d6ac9) | diff |
| --- | --- | --- | --- |
| `ml` | 1234 | 1237 | +3 |

#### async chunks size
| id | [before](https://github.com/elastic/kibana/commit/6f2dc107275b8a2d264cb9148c70302c7d2d8275) | [after](https://github.com/elastic/kibana/commit/21bb6cb31d31959d096b318b15f2426c840d6ac9) | diff |
| --- | --- | --- | --- |
| `ml` | 6.6MB | 6.6MB | +17.9KB |

### History
* :green_heart: [Build #83747](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/) succeeded 3c8972a6016d952a4380f5113ff28f50cb7f16ff

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/","number":"83747","commit":"3c8972a6016d952a4380f5113ff28f50cb7f16ff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85403/","number":"85403","commit":"21bb6cb31d31959d096b318b15f2426c840d6ac9"}]}
PIPELINE-->

peteharverson commented on x-pack/plugins/ml/public/application/services/results_service/results_service.js:
```diff
@@ -1300,6 +1307,14 @@ export function resultsServiceProvider(mlApiServices) {
+          mustCriteria.push({
+            term: {
+              function_description:
+                actualPlotFunctionIfMetric === 'avg' ? 'mean' : actualPlotFunctionIfMetric,
```
Can you use the enums out of `common/constants/aggregation_types` here?

peteharverson commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer.js:
```diff
@@ -378,6 +395,71 @@ export class TimeSeriesExplorer extends React.Component {
+   * @param {Array} entities - Entity controls configuration
+   * @param {Object} searchTerm - Search term for partition, e.g. { partition_field: 'partition' }
+   */
+  loadEntityValues = async (entities, searchTerm = {}) => {
```
Does this need removing in favor of the changes made in #81510 ?

qn895 replied:
Deleted here https://github.com/elastic/kibana/pull/81662/commits/a583988eba972e5a1e395aff7c7f613464407e10

peteharverson commented on x-pack/plugins/ml/server/models/results_service/results_service.ts:
```diff
@@ -106,6 +107,14 @@ export function resultsServiceProvider(client: IScopedClusterClient) {
+    if (actualPlotFunction !== undefined) {
+      boolCriteria.push({
+        term: {
+          function_description: actualPlotFunction === 'avg' ? 'mean' : actualPlotFunction,
```
Can you use the enums out of `common/constants/aggregation_types` here?

peteharverson commented on x-pack/plugins/ml/server/models/results_service/results_service.ts:
```diff
@@ -57,7 +57,8 @@ export function resultsServiceProvider(client: IScopedClusterClient) {
     maxExamples: number = DEFAULT_MAX_EXAMPLES,
-    influencersFilterQuery: any
+    influencersFilterQuery?: any,
+    actualPlotFunction?: string | undefined
```
Maybe this should be renamed to `functionDescription` to match the name of the property in the records?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/cb2429e192079f4e9a1cfc0e401a2a9544d19f23

peteharverson commented on x-pack/plugins/ml/server/routes/results_service.ts:
```diff
@@ -37,6 +37,7 @@ function getAnomaliesTableData(client: IScopedClusterClient, payload: any) {
     maxRecords,
     maxExamples,
     influencersFilterQuery,
+    actualPlotFunction,
```
As before, maybe this should be renamed to `functionDescription` to match the name of the property in the records?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/cb2429e192079f4e9a1cfc0e401a2a9544d19f23

peteharverson commented on x-pack/plugins/ml/server/routes/schemas/results_service_schema.ts:
```diff
@@ -26,6 +26,7 @@ export const anomaliesTableDataSchema = schema.object({
   maxRecords: schema.number(),
   maxExamples: schema.maybe(schema.number()),
   influencersFilterQuery: schema.maybe(schema.any()),
+  actualPlotFunction: schema.maybe(schema.nullable(schema.string())),
```
As above, maybe this should be renamed to `functionDescription` to match the name of the property in the records?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/cb2429e192079f4e9a1cfc0e401a2a9544d19f23

peteharverson commented on x-pack/plugins/ml/public/application/services/ml_api_service/results.ts:
```diff
@@ -24,8 +24,9 @@ export const resultsApiProvider = (httpService: HttpService) => ({
-    influencersFilterQuery: any
+    maxExamples?: number,
+    influencersFilterQuery?: any,
+    actualPlotFunction?: string | undefined
```
Maybe this should be renamed to `functionDescription` to match the name of the property in the records?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/cb2429e192079f4e9a1cfc0e401a2a9544d19f23

peteharverson commented:
Coming into the view from the jobs list, I see the function dropdown disappear after selecting a partitioning field:

![metric_chart_control_bug](https://user-images.githubusercontent.com/7405507/97987375-11456e80-1dd3-11eb-95a0-bffb64dc6eae.gif)

peteharverson commented:
Noticed some odd behavior in the chart when using the metric function, which is happening here and also on master. For these anomalies which are in `min`, the chart data seems to be plotted always with `avg`. Here the anomaly should be plotted with a value of 53.6:

![image](https://user-images.githubusercontent.com/7405507/97989164-0b04c180-1dd6-11eb-890d-eae514ed4241.png)

```
  "analysis_config": {
    "bucket_span": "15m",
    "detectors": [
      {
        "detector_description": "metric(responsetime) partitionfield=airline",
        "function": "metric",
        "field_name": "responsetime",
        "partition_field_name": "airline",
        "detector_index": 0
      }
    ],
    "influencers": [
      "airline"
    ]
  },
  "analysis_limits": {
    "model_memory_limit": "13mb",
    "categorization_examples_limit": 4
  },
  "data_description": {
    "time_field": "@timestamp",
    "time_format": "epoch_ms"
  },
  "model_plot_config": {
    "enabled": true,
    "annotations_enabled": true
  },
```

qn895 commented:
> Coming into the view from the jobs list, I see the function dropdown disappear after selecting a partitioning field

@peteharverson I have added a fix for this issue. The dropdown should now work with or without the partitioning and should be updated.

> Noticed some odd behavior in the chart when using the metric function, which is happening here and also on master. For these anomalies which are in min, the chart data seems to be plotted always with avg.

I also added a fix for this one in https://github.com/elastic/kibana/pull/81662/commits/ea8d71fd4e798dde7b668bb084d1bc079631db1a. Now it will 1) get the anomaly records based on the selected function description, and 2) plot the chart point with the actual data point.

<img width="1093" alt="Screen Shot 2020-11-03 at 14 15 08" src="https://user-images.githubusercontent.com/43350163/98036384-8f0f7700-1ddf-11eb-82c4-e9a2a029e442.png">
<img width="1093" alt="Screen Shot 2020-11-03 at 14 15 16" src="https://user-images.githubusercontent.com/43350163/98036387-920a6780-1ddf-11eb-94a6-e4aec0005b47.png">

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85635/)
* Commit: f6948683bff5517db3b5907e600ec8fcaab2da84

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [before](https://github.com/elastic/kibana/commit/ac331482aa09d898728975ea7bc9fbc676594901) | [after](https://github.com/elastic/kibana/commit/f6948683bff5517db3b5907e600ec8fcaab2da84) | diff |
| --- | --- | --- | --- |
| `ml` | 1237 | 1241 | +4 |

#### async chunks size
| id | [before](https://github.com/elastic/kibana/commit/ac331482aa09d898728975ea7bc9fbc676594901) | [after](https://github.com/elastic/kibana/commit/f6948683bff5517db3b5907e600ec8fcaab2da84) | diff |
| --- | --- | --- | --- |
| `ml` | 6.6MB | 6.6MB | +16.7KB |

### History
* :green_heart: [Build #85403](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85403/) succeeded 21bb6cb31d31959d096b318b15f2426c840d6ac9
* :green_heart: [Build #83747](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/) succeeded 3c8972a6016d952a4380f5113ff28f50cb7f16ff

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/","number":"83747","commit":"3c8972a6016d952a4380f5113ff28f50cb7f16ff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85403/","number":"85403","commit":"21bb6cb31d31959d096b318b15f2426c840d6ac9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85635/","number":"85635","commit":"f6948683bff5517db3b5907e600ec8fcaab2da84"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85650/)
* Commit: 09fe186c1cc592c309d38466571a7441b36f330e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [before](https://github.com/elastic/kibana/commit/ac331482aa09d898728975ea7bc9fbc676594901) | [after](https://github.com/elastic/kibana/commit/09fe186c1cc592c309d38466571a7441b36f330e) | diff |
| --- | --- | --- | --- |
| `ml` | 1237 | 1241 | +4 |

#### async chunks size
| id | [before](https://github.com/elastic/kibana/commit/ac331482aa09d898728975ea7bc9fbc676594901) | [after](https://github.com/elastic/kibana/commit/09fe186c1cc592c309d38466571a7441b36f330e) | diff |
| --- | --- | --- | --- |
| `ml` | 6.6MB | 6.6MB | +16.0KB |

### History
* :green_heart: [Build #85635](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85635/) succeeded f6948683bff5517db3b5907e600ec8fcaab2da84
* :green_heart: [Build #85403](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85403/) succeeded 21bb6cb31d31959d096b318b15f2426c840d6ac9
* :green_heart: [Build #83747](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/) succeeded 3c8972a6016d952a4380f5113ff28f50cb7f16ff

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/","number":"83747","commit":"3c8972a6016d952a4380f5113ff28f50cb7f16ff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85403/","number":"85403","commit":"21bb6cb31d31959d096b318b15f2426c840d6ac9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85635/","number":"85635","commit":"f6948683bff5517db3b5907e600ec8fcaab2da84"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85650/","number":"85650","commit":"09fe186c1cc592c309d38466571a7441b36f330e"}]}
PIPELINE-->

darnautov commented on x-pack/plugins/ml/public/application/timeseriesexplorer/get_function_description.ts:
```diff
@@ -0,0 +1,58 @@
+      return highestScoringAnomaly?.function_description;
+    }
+  } catch (error) {
+    getToastNotificationService().displayErrorToast(
```
can you please use a service from the Kibana context?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/5350e3da75c6812e47882cd4818e3ae98343c08c

darnautov commented on x-pack/plugins/ml/public/application/timeseriesexplorer/components/plot_function_controls/plot_function_controls.tsx:
```diff
@@ -0,0 +1,54 @@
+}) => {
+  if (functionDescription === undefined) return null;
+  return (
+    <EuiFlexItem style={{ textAlign: 'right' }} grow={false}>
```
to avoid inline styles, you can use [EuiText](https://elastic.github.io/eui/#/display/text) with `textAlign` prop

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/a7e2482b474068eec863111aab585bbd84b75e40

darnautov commented on x-pack/plugins/ml/public/application/timeseriesexplorer/components/plot_function_controls/plot_function_controls.tsx:
```diff
@@ -0,0 +1,54 @@
+          options={plotByFunctionOptions}
+          value={functionDescription}
+          onChange={(e) => setFunctionDescription(e.target.value)}
+          aria-label="Pick function to plot by (min, max, or average) if metric function"
```
`aria-label` also requires internationalization support

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/a7e2482b474068eec863111aab585bbd84b75e40

peteharverson commented:
Found what looks like an issue with setting the bounds on the focus chart:

![image](https://user-images.githubusercontent.com/7405507/98121465-3a373380-1ea7-11eb-83ca-5cb4cba39116.png)

Config is:
```
 "analysis_config": {
    "bucket_span": "15m",
    "detectors": [
      {
        "detector_description": "metric(responsetime) partitionfield=airline",
        "function": "metric",
        "field_name": "responsetime",
        "partition_field_name": "airline",
        "detector_index": 0
      }
    ],
    "influencers": [
      "airline"
    ]
  },
```

qn895 commented:
@peteharverson I have fixed the two issues in my latest changes. The bounds should be now correct, and the mini swimlanes should should up correctly for when `function = mean` as well.

<img width="1208" alt="Screen Shot 2020-11-04 at 11 13 53" src="https://user-images.githubusercontent.com/43350163/98143592-e7ea1880-1e8e-11eb-9084-0fe43cab858a.png">
<img width="1208" alt="Screen Shot 2020-11-04 at 11 14 01" src="https://user-images.githubusercontent.com/43350163/98143599-ea4c7280-1e8e-11eb-8950-627b59417b55.png">

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85865/)
* Commit: 595b8b36dc4770f37160b19173d6614c09933759

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [before](https://github.com/elastic/kibana/commit/89887984d8b2415a7cc0a746d3b7e58108cade96) | [after](https://github.com/elastic/kibana/commit/595b8b36dc4770f37160b19173d6614c09933759) | diff |
| --- | --- | --- | --- |
| `ml` | 1238 | 1242 | +4 |

#### async chunks size
| id | [before](https://github.com/elastic/kibana/commit/89887984d8b2415a7cc0a746d3b7e58108cade96) | [after](https://github.com/elastic/kibana/commit/595b8b36dc4770f37160b19173d6614c09933759) | diff |
| --- | --- | --- | --- |
| `ml` | 6.6MB | 6.7MB | +17.0KB |

### History
* :green_heart: [Build #85650](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85650/) succeeded 09fe186c1cc592c309d38466571a7441b36f330e
* :green_heart: [Build #85635](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85635/) succeeded f6948683bff5517db3b5907e600ec8fcaab2da84
* :green_heart: [Build #85403](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85403/) succeeded 21bb6cb31d31959d096b318b15f2426c840d6ac9
* :green_heart: [Build #83747](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/) succeeded 3c8972a6016d952a4380f5113ff28f50cb7f16ff

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/","number":"83747","commit":"3c8972a6016d952a4380f5113ff28f50cb7f16ff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85403/","number":"85403","commit":"21bb6cb31d31959d096b318b15f2426c840d6ac9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85635/","number":"85635","commit":"f6948683bff5517db3b5907e600ec8fcaab2da84"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85650/","number":"85650","commit":"09fe186c1cc592c309d38466571a7441b36f330e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85865/","number":"85865","commit":"595b8b36dc4770f37160b19173d6614c09933759"}]}
PIPELINE-->

walterra commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer.js:
```diff
@@ -392,13 +426,11 @@ export class TimeSeriesExplorer extends React.Component {
-    const currentSelectedJob = mlJobService.getJob(selectedJobId);
+    const functionToPlotByIfMetric = aggregationTypeTransform.toES(functionDescription);
 
-    if (currentSelectedJob === undefined) {
-      return;
-    }
```
Can you explain why this check is no longer necessary?

qn895 replied:
Thanks for catching that. This was removed by accident. I have revert the change in https://github.com/elastic/kibana/pull/81662/commits/64ecedbd1dccaa5b916c4e81538729a17dfb3afb

walterra commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer.js:
```diff
@@ -810,14 +843,23 @@ export class TimeSeriesExplorer extends React.Component {
+      previousProps.selectedDetectorIndex !== this.props.selectedDetectorIndex ||
+      !isEqual(previousProps.selectedEntities, this.props.selectedEntities)
+    ) {
+      this.getFunctionDescription();
```
Could we run into problems with a race condition since this is async function?
(Just trying to understand the implications here, since previously only relying on a change `previousProps` did make sure we check against from changes coming from outside the components - wondering if with including `perviousState` and then again triggering async state functions if we might run into render loops)

walterra reviewed (commented):
Added some questions, one is about risk of render looks or race conditions.

peteharverson commented on x-pack/plugins/ml/public/application/services/results_service/result_service_rx.ts:
```diff
@@ -347,9 +348,10 @@ export function resultsServiceRxProvider(mlApiServices: MlApiServices) {
+      earliestMs: number | null,
+      latestMs: number | null,
+      maxResults: number | undefined,
+      functionDescription?: string | undefined
```
As noted above, you should be able to remove ` | undefined` because of the `?`.

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/64ecedbd1dccaa5b916c4e81538729a17dfb3afb

peteharverson commented on x-pack/plugins/ml/public/application/explorer/explorer_charts/explorer_chart_config_builder.js:
```diff
@@ -48,6 +49,10 @@ export function buildConfig(record) {
   // define the metric series to be plotted.
   config.entityFields = getEntityFieldList(record);
 
+  if (record.function === 'metric') {
```
Use `ML_JOB_AGGREGATION.METRIC` from `common/constants/aggregation_types`

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/64ecedbd1dccaa5b916c4e81538729a17dfb3afb

peteharverson commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer_utils/timeseriesexplorer_utils.js:
```diff
@@ -142,6 +143,8 @@ export function processDataForFocusAnomalies(
     // Look for a chart point with the same time as the record.
     // If none found, find closest time in chartData set.
     const recordTime = record[TIME_FIELD_NAME];
+    if (record.function === 'metric' && record.function_description !== functionDescription) return;
```
Use `ML_JOB_AGGREGATION.METRIC` from `common/constants/aggregation_types`

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/64ecedbd1dccaa5b916c4e81538729a17dfb3afb

peteharverson commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer_utils/timeseriesexplorer_utils.js:
```diff
@@ -160,6 +163,10 @@ export function processDataForFocusAnomalies(
             chartPoint.value = record.actual;
           }
 
+          if (record.function === 'metric') {
```
Use `ML_JOB_AGGREGATION.METRIC` from `common/constants/aggregation_types`

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/64ecedbd1dccaa5b916c4e81538729a17dfb3afb

peteharverson commented on x-pack/plugins/ml/public/application/timeseriesexplorer/get_function_description.ts:
```diff
@@ -0,0 +1,57 @@
+) => {
+  // if the detector's function is metric, fetch the highest scoring anomaly record
+  // and set to plot the function_description (avg/min/max) of that record by default
+  if (selectedJob?.analysis_config?.detectors[selectedDetectorIndex]?.function !== 'metric') return;
```
Use `ML_JOB_AGGREGATION.METRIC` from `common/constants/aggregation_types`

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/81662/commits/64ecedbd1dccaa5b916c4e81538729a17dfb3afb

peteharverson reviewed (approved):
Tested latest edits and LGTM.
Just left some minor comments on the code.

qn895 commented on x-pack/plugins/ml/public/application/timeseriesexplorer/components/plot_function_controls/plot_function_controls.tsx:
```diff
@@ -0,0 +1,56 @@
+import { EuiFlexItem, EuiFormRow, EuiSelect } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+
+const plotByFunctionOptions = [
```
Since these are not strict terms that are excluded in the translation guide, I'll keep these translations here.

qn895 commented on x-pack/plugins/ml/public/application/services/ml_api_service/results.ts:
```diff
@@ -24,8 +24,9 @@ export const resultsApiProvider = (httpService: HttpService) => ({
-    influencersFilterQuery: any
+    maxExamples?: number,
+    influencersFilterQuery?: any,
+    functionDescription?: string | undefined
```
Updated here https://github.com/elastic/kibana/pull/81662/commits/64ecedbd1dccaa5b916c4e81538729a17dfb3afb

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/86636/)
* Commit: df26340ce0f8f2c54d948548ea9af81b692444ec

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [before](https://github.com/elastic/kibana/commit/8560b2dcc00f3992b43b75daf844ce7834678ac8) | [after](https://github.com/elastic/kibana/commit/df26340ce0f8f2c54d948548ea9af81b692444ec) | diff |
| --- | --- | --- | --- |
| `ml` | 1238 | 1242 | +4 |

#### async chunks size
| id | [before](https://github.com/elastic/kibana/commit/8560b2dcc00f3992b43b75daf844ce7834678ac8) | [after](https://github.com/elastic/kibana/commit/df26340ce0f8f2c54d948548ea9af81b692444ec) | diff |
| --- | --- | --- | --- |
| `ml` | 6.6MB | 6.7MB | +17.3KB |

### History
* :green_heart: [Build #85865](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85865/) succeeded 595b8b36dc4770f37160b19173d6614c09933759
* :green_heart: [Build #85650](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85650/) succeeded 09fe186c1cc592c309d38466571a7441b36f330e
* :green_heart: [Build #85635](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85635/) succeeded f6948683bff5517db3b5907e600ec8fcaab2da84
* :green_heart: [Build #85403](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85403/) succeeded 21bb6cb31d31959d096b318b15f2426c840d6ac9
* :green_heart: [Build #83747](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/) succeeded 3c8972a6016d952a4380f5113ff28f50cb7f16ff

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83747/","number":"83747","commit":"3c8972a6016d952a4380f5113ff28f50cb7f16ff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85403/","number":"85403","commit":"21bb6cb31d31959d096b318b15f2426c840d6ac9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85635/","number":"85635","commit":"f6948683bff5517db3b5907e600ec8fcaab2da84"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85650/","number":"85650","commit":"09fe186c1cc592c309d38466571a7441b36f330e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/85865/","number":"85865","commit":"595b8b36dc4770f37160b19173d6614c09933759"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/86636/","number":"86636","commit":"df26340ce0f8f2c54d948548ea9af81b692444ec"}]}
PIPELINE-->

darnautov reviewed (approved):
LGTM

walterra reviewed (approved):
LGTM, we can think of ways to improve `componentDidUpdate()` in the single metric viewer component in a follow up.
